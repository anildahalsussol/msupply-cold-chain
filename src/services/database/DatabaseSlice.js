import { createSlice, combineReducers } from '@reduxjs/toolkit';
import { REDUCER_SHAPE } from '~constants';

export const { actions: SensorsActions, reducer: SensorsReducer } = createSlice({
  name: REDUCER_SHAPE.SENSORS,
  initialState: { sensorIds: [], byId: {} },
  reducers: {
    setSensors: {
      reducer: (_, action) => {
        const { payload } = action;
        const { sensors } = payload;

        const byId = sensors.reduce((acc, sensor) => {
          const { id } = sensor;
          acc[id] = sensor;
          return acc;
        }, {});

        const sensorIds = sensors.map(({ id }) => id);

        return { byId, sensorIds };
      },
      prepare: sensors => ({ payload: { sensors } }),
    },

    saveSensors: {
      reducer: () => {},
      prepare: sensors => ({ payload: { sensors } }),
    },
  },
});

export const { actions: TemperatureLogActions, reducer: TemperatureLogReducer } = createSlice({
  name: REDUCER_SHAPE.TEMPERATURE_LOGS,
  initialState: { bySensorId: {}, fromDate: null, toDate: null },
  reducers: {
    saveTemperatureLogs: {
      reducer: () => {},
      prepare: temperatureLogs => ({ payload: { temperatureLogs } }),
    },
    saveSensorLogs: {
      reducer: () => {},
      prepare: (data, macAddress) => ({ payload: { data, macAddress } }),
    },
    setTemperatureLogs: {
      prepare: temperatureLogs => ({ payload: { temperatureLogs } }),
    },
    createTemperatureLogs: {
      prepare: macAddress => ({ payload: { macAddress } }),
      reducer: () => {},
    },
    savedTemperatureLogs: {
      prepare: (macAddress, logs) => ({ payload: { macAddress, logs } }),
      reducer: () => {},
    },
    createBreaches: {
      prepare: macAddress => ({ payload: { macAddress } }),
      reducer: () => {},
    },
  },
});

export const DatabaseReducer = combineReducers({
  sensors: SensorsReducer,
  temperatureLogs: TemperatureLogReducer,
});
