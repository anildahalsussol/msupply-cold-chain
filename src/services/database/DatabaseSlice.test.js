import { SensorsActions, SensorsReducer, TemperatureLogActions } from './DatabaseSlice';

describe('SensorsActions', () => {
  it('setSensors', () => {
    const sensors = [{ id: 1 }];
    expect(SensorsActions.setSensors(sensors)).toMatchSnapshot();
  });
  it('saveSensors', () => {
    const sensors = [{ id: 1 }];
    expect(SensorsActions.saveSensors(sensors)).toMatchSnapshot();
  });
});

describe('TemperatureLogActions', () => {
  it('saveTemperatureLogs', () => {
    const logs = [{ id: 1 }];
    expect(TemperatureLogActions.saveTemperatureLogs(logs)).toMatchSnapshot();
  });
  it('saveSensorLogs', () => {
    const sensorLogs = [{ id: 1 }];
    const mac = 'a';
    expect(TemperatureLogActions.saveSensorLogs(sensorLogs, mac)).toMatchSnapshot();
  });
  it('setTemperatureLogs', () => {
    const logs = [{ id: 1 }];
    expect(TemperatureLogActions.setTemperatureLogs(logs)).toMatchSnapshot();
  });
  it('createTemperatureLogs', () => {
    const mac = 'a';
    expect(TemperatureLogActions.createTemperatureLogs(mac)).toMatchSnapshot();
  });
  it('savedTemperatureLogs', () => {
    const mac = 'a';
    const logs = [{ id: 1 }];
    expect(TemperatureLogActions.savedTemperatureLogs(mac, logs)).toMatchSnapshot();
  });
  it('createBreaches', () => {
    const mac = 'a';
    expect(TemperatureLogActions.createBreaches(mac)).toMatchSnapshot();
  });
  it('savedBreaches', () => {
    const created = [];
    const ended = [];
    expect(TemperatureLogActions.savedBreaches(created, ended)).toMatchSnapshot();
  });
});

describe('SensorsReducer', () => {
  it('setSensors', () => {
    const sensors = [{ id: 1 }];
    const initialState = { sensorIds: [], byId: {} };
    expect(SensorsReducer(initialState, SensorsActions.setSensors(sensors))).toEqual({
      sensorIds: [1],
      byId: { 1: { id: 1 } },
    });
  });
  it('saveSensors', () => {
    const sensors = [{ id: 1 }];
    const initialState = { sensorIds: [], byId: {} };
    expect(SensorsReducer(initialState, SensorsActions.saveSensors(sensors))).toEqual(initialState);
  });
});

describe('TemperatureLogReducer', () => {
  it('saveTemperatureLogs', () => {
    const initialState = { bySensorId: {}, fromDate: null, toDate: null };
    expect(SensorsReducer(initialState, TemperatureLogActions.saveTemperatureLogs())).toEqual(
      initialState
    );
  });
  it('saveSensorLogs', () => {
    const initialState = { bySensorId: {}, fromDate: null, toDate: null };
    expect(SensorsReducer(initialState, TemperatureLogActions.saveSensorLogs())).toEqual(
      initialState
    );
  });
  it('setTemperatureLogs', () => {
    const initialState = { bySensorId: {}, fromDate: null, toDate: null };
    expect(SensorsReducer(initialState, TemperatureLogActions.setTemperatureLogs())).toEqual(
      initialState
    );
  });
  it('createTemperatureLogs', () => {
    const initialState = { bySensorId: {}, fromDate: null, toDate: null };
    const mac = 'a';
    expect(SensorsReducer(initialState, TemperatureLogActions.setTemperatureLogs(mac))).toEqual(
      initialState
    );
  });
  it('savedTemperatureLogs', () => {
    const initialState = { bySensorId: {}, fromDate: null, toDate: null };
    expect(SensorsReducer(initialState, TemperatureLogActions.savedTemperatureLogs())).toEqual(
      initialState
    );
  });
  it('createBreaches', () => {
    const initialState = { bySensorId: {}, fromDate: null, toDate: null };
    expect(SensorsReducer(initialState, TemperatureLogActions.createBreaches())).toEqual(
      initialState
    );
  });
  it('savedBreaches', () => {
    const initialState = { bySensorId: {}, fromDate: null, toDate: null };
    expect(SensorsReducer(initialState, TemperatureLogActions.savedBreaches())).toEqual(
      initialState
    );
  });
});
