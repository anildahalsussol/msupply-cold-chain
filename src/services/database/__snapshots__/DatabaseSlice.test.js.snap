// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SensorsActions saveSensors 1`] = `
Object {
  "payload": Object {
    "sensors": Array [
      Object {
        "id": 1,
      },
    ],
  },
  "type": "sensors/saveSensors",
}
`;

exports[`SensorsActions setSensors 1`] = `
Object {
  "payload": Object {
    "sensors": Array [
      Object {
        "id": 1,
      },
    ],
  },
  "type": "sensors/setSensors",
}
`;

exports[`TemperatureLogActions createBreaches 1`] = `
Object {
  "payload": Object {
    "macAddress": "a",
  },
  "type": "temperature_logs/createBreaches",
}
`;

exports[`TemperatureLogActions createTemperatureLogs 1`] = `
Object {
  "payload": Object {
    "macAddress": "a",
  },
  "type": "temperature_logs/createTemperatureLogs",
}
`;

exports[`TemperatureLogActions saveSensorLogs 1`] = `
Object {
  "payload": Object {
    "data": Array [
      Object {
        "id": 1,
      },
    ],
    "macAddress": "a",
  },
  "type": "temperature_logs/saveSensorLogs",
}
`;

exports[`TemperatureLogActions saveTemperatureLogs 1`] = `
Object {
  "payload": Object {
    "temperatureLogs": Array [
      Object {
        "id": 1,
      },
    ],
  },
  "type": "temperature_logs/saveTemperatureLogs",
}
`;

exports[`TemperatureLogActions savedBreaches 1`] = `
Object {
  "payload": Object {
    "createdBreaches": Array [],
    "endedBreaches": Array [],
  },
  "type": "temperature_logs/savedBreaches",
}
`;

exports[`TemperatureLogActions savedTemperatureLogs 1`] = `
Object {
  "payload": Object {
    "logs": Array [
      Object {
        "id": 1,
      },
    ],
    "macAddress": "a",
  },
  "type": "temperature_logs/savedTemperatureLogs",
}
`;

exports[`TemperatureLogActions setTemperatureLogs 1`] = `
Object {
  "payload": Object {
    "temperatureLogs": Array [
      Object {
        "id": 1,
      },
    ],
  },
  "type": "temperature_logs/setTemperatureLogs",
}
`;
